core/src/main/scala/org/apache/spark/executor/ProcfsBasedSystems.scala

```diff
@@ -0,0 +1,211 @@
+import scala.collection.mutable.Queue
+
+// Some of the ideas here are taken from the ProcfsBasedProcessTree class in hadoop
+// project.
+class ProcfsBasedSystems  extends ProcessTreeMetrics {
+  val procfsDir = "/proc/"
+  var isAvailable: Boolean = isItProcfsBased
+  val pid: Int = computePid()
+  val ptree: scala.collection.mutable.Map[ Int, Set[Int]] =
+    scala.collection.mutable.Map[ Int, Set[Int]]()
+  val PROCFS_STAT_FILE = "stat"
+
+
+  def isItProcfsBased: Boolean = {
+    val testing = sys.env.contains("SPARK_TESTING") || sys.props.contains("spark.testing")
+    if (testing) {
+      return true
+    }
+    try {
+      if (!Files.exists(Paths.get(procfsDir))) {
+        return false
+      }
+    }
+    catch {
+      case f: FileNotFoundException => return false
+    }
+    true
+  }
+
+
+  def computePid(): Int = {
+    if (!isAvailable) {
+      return -1;
+    }
+    val cmd = Array("bash", "-c", "echo $PPID")
+    val length = 10
+    var out: Array[Byte] = Array.fill[Byte](length)(0)
+    Runtime.getRuntime.exec(cmd).getInputStream.read(out)
+    val pid = Integer.parseInt(new String(out, "UTF-8").trim)
+    return pid;
+  }
+
+
+  def createProcessTree(): Unit = {
+    if (!isAvailable) {
+      return
+    }
+    val queue: Queue[Int] = new Queue[Int]()
+    queue += pid
+    while( !queue.isEmpty ) {
+      val p = queue.dequeue()
+      val c = getChildPIds(p)
+      if(!c.isEmpty) {
+        queue ++= c
+        ptree += (p -> c.toSet)
+      }
+      else {
+        ptree += (p -> Set[Int]())
+      }
+    }
+  }
+
+
+  def updateProcessTree(): Unit = {
+    if (!isAvailable) {
+      return
+    }
+    val queue: Queue[Int] = new Queue[Int]()
+    queue += pid
+    while( !queue.isEmpty ) {
+      val p = queue.dequeue()
+      val c = getChildPIds(p)
+      if(!c.isEmpty) {
+        queue ++= c
+        val preChildren = ptree.get(p)
+        preChildren match {
+          case Some(children) => if (!c.toSet.equals(children)) {
+            val diff: Set[Int] = children -- c.toSet
+            ptree.update(p, c.toSet )
+            diff.foreach(ptree.remove(_))
+          }
+          case None => ptree.update(p, c.toSet )
+        }
+      }
+      else {
+        ptree.update(p, Set[Int]())
+      }
+    }
+  }
+
+
+  /**
+   * Hadoop ProcfsBasedProcessTree class used regex and pattern matching to retrive the memory
+   * info. I tried that but found it not correct during tests, so I used normal string analysis
+   * instead. The computation of RSS and Vmem are based on proc(5):
+   * http://man7.org/linux/man-pages/man5/proc.5.html
+   */
+  def getProcessInfo(pid: Int): String = {
+    try {
+      val pidDir: File = new File(procfsDir, pid.toString)
+      val fReader = new InputStreamReader(
+        new FileInputStream(
+          new File(pidDir, PROCFS_STAT_FILE)), Charset.forName("UTF-8"))
+      val in: BufferedReader = new BufferedReader(fReader)
+      val procInfo = in.readLine
```

Comment:
Can you please add a comment why we are reading just first line?

Reply:
This is what hadoop's ProcfsBasedProcessTree is doing as well. I wasn't able to find a reference, but in my testing also reading just oneline was enough.